File: db4oj.tests/src/com/db4o/db4ounit/common/querying/CascadedDeleteUpdate.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.db4ounit.common.querying;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class CascadedDeleteUpdate extends AbstractDb4oTestCase {

	public static class CduHelper {
		Object parent1;
		Object parent2;
	}
	
	public Object child;
	
	protected void configure(Configuration config) {
		config.objectClass(this).cascadeOnDelete(true);
		config.objectClass(CduHelper.class).cascadeOnDelete(true);
	}
	
	protected void store() {
		CascadedDeleteUpdate cdu1 = new CascadedDeleteUpdate();
		CascadedDeleteUpdate cdu2 = new CascadedDeleteUpdate();
		
		CduHelper helper = new CduHelper();
		helper.parent1 = cdu1;
		helper.parent2 = cdu2;
		cdu1.child = helper; 
		cdu2.child = helper;
		
		db().set(cdu1);
		db().set(cdu2);
		db().set(cdu1);
		db().set(cdu2);
	}
	
	public void test(){
		Query q = newQuery();
		q.constrain(this.getClass());
		ObjectSet objectSet = q.execute();
		
		while(objectSet.hasNext()){
			CascadedDeleteUpdate cdu = (CascadedDeleteUpdate) objectSet.next();
			Assert.isNotNull(cdu.child);
		}
	}
}

File: db4on/test/src/com/db4o/db4ounit/common/querying/CascadedDeleteUpdate.cs
namespace com.db4o.db4ounit.common.querying
{
	public class CascadedDeleteUpdate : Db4oUnit.Extensions.AbstractDb4oTestCase
	{
		public class ParentItem
		{
			public object child;
		}

		public class ChildItem
		{
			public object parent1;

			public object parent2;
		}

		public static void Main(string[] arguments)
		{
			new com.db4o.db4ounit.common.querying.CascadedDeleteUpdate().RunClientServer();
		}

		protected override void Configure(com.db4o.config.Configuration config)
		{
			config.ObjectClass(typeof(com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem)
				).CascadeOnDelete(true);
		}

		protected override void Store()
		{
			com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem parentItem1 = new 
				com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem();
			com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem parentItem2 = new 
				com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem();
			com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ChildItem child = new com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ChildItem
				();
			child.parent1 = parentItem1;
			child.parent2 = parentItem2;
			parentItem1.child = child;
			parentItem2.child = child;
			Db().Set(parentItem1);
		}

		public virtual void TestAllObjectStored()
		{
			AssertAllObjectStored();
		}

		public virtual void TestUpdate()
		{
			com.db4o.query.Query q = NewQuery(typeof(com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem)
				);
			com.db4o.ObjectSet objectSet = q.Execute();
			while (objectSet.HasNext())
			{
				Db().Set(objectSet.Next());
			}
			Db().Commit();
			AssertAllObjectStored();
		}

		private void AssertAllObjectStored()
		{
			Reopen();
			com.db4o.query.Query q = NewQuery(typeof(com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem)
				);
			com.db4o.ObjectSet objectSet = q.Execute();
			while (objectSet.HasNext())
			{
				com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem parentItem = (com.db4o.db4ounit.common.querying.CascadedDeleteUpdate.ParentItem
					)objectSet.Next();
				Db().Refresh(parentItem, 3);
				Db4oUnit.Assert.IsNotNull(parentItem.child);
			}
		}
	}
}
